src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpDeclareAsNullableCodeFixProvider.cs

```diff
@@ -33,66 +38,96 @@ public CSharpDeclareAsNullableCodeFixProvider()
+                GetEquivalenceKey(node, model)),
                 context.Diagnostics);
         }
 
-        private string GetEquivalenceKey(SyntaxNode node)
+        private string GetEquivalenceKey(SyntaxNode node, SemanticModel model)
         {
-            return node.IsKind(SyntaxKind.ConditionalAccessExpression) ? IsConditionalOperatorEquivalenceKey : IsOtherEquivalenceKey;
+            return IsRemoteApiUsage(node, model) ? AssigningNullLiteralRemotelyEquivalenceKey :
+                node.IsKind(SyntaxKind.ConditionalAccessExpression) ? ConditionalOperatorEquivalenceKey :
+                AssigningNullLiteralLocallyEquivalenceKey;
+
+            static bool IsRemoteApiUsage(SyntaxNode node, SemanticModel model)
+            {
+                if (node.IsParentKind(SyntaxKind.Argument))
+                {
+                    // M(null) could be used in a test
+                    return true;
+                }
+
+                if (node.Parent is AssignmentExpressionSyntax assignment)
+                {
+                    var symbol = model.GetSymbolInfo(assignment.Left).Symbol;
+                    if (symbol is IFieldSymbol)
+                    {
+                        // x.field could be used in a test
+                        return true;
+                    }
+                    else if (symbol is IPropertySymbol)
+                    {
+                        // x.Property could be used in a test
+                        return true;
+                    }
+                }
+
+                return false;
+            }
         }
 
-        protected override Task FixAllAsync(
+        protected override async Task FixAllAsync(
             Document document, ImmutableArray<Diagnostic> diagnostics,
             SyntaxEditor editor, CancellationToken cancellationToken)
         {
             // a method can have multiple `return null;` statements, but we should only fix its return type once
             using var _ = PooledHashSet<TypeSyntax>.GetInstance(out var alreadyHandled);
 
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             foreach (var diagnostic in diagnostics)
             {
                 var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
-                MakeDeclarationNullable(editor, node, alreadyHandled);
+                MakeDeclarationNullable(editor, node, alreadyHandled, model);
             }
-
-            return Task.CompletedTask;
         }
 
         protected override bool IncludeDiagnosticDuringFixAll(Diagnostic diagnostic, Document document, string equivalenceKey, CancellationToken cancellationToken)
         {
             var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
-            return equivalenceKey == GetEquivalenceKey(node);
+            var model = document.GetSemanticModelAsync(cancellationToken).GetAwaiter().GetResult();
```

Comment:
:memo: I don't really like doing this. Should I make `IncludeDiagnosticDuringFixAll` async too?

Reply:
the intended design is that the diagnostic contains enough info it to make this determination.  otherwise, you can push this check into the actual FixAll logic (that is already async).

Reply:
interesting.  in this case you have teh same compiler diagnostic, but for very different scenarios.  i don't think we've had to deal with that before.   I'm not a fan of making this async as i don't think we want to get the semantic-model per document.

instead, the caller should get the semantic model and the pass that into this helper.  to keep churn low you can add in a new virtual that takes in the SemanticModel as well, it will normally call the existing helper.  Hover, this class can then override that new virtual (and ignore the other one).